### DISCUSSION TITLE
Dynamic Attributes and Integer Increments

### DISCUSSION THREAD

**Discussion by shortjared** [tywalch/electrodb#235] [Q&A]
Hello,

I'm trying to figure out if it would be possible to model a dynamic map and do integer increments on values in the map. Most my attempts so far have failed with DynamoDB saying the update expression is invalid, since electro puts things in `ADD #attr :attr` instead of something like `SET #attr.#prop = if_not_exists(#attr.#prop, :defaultZero) + :incr`

There is the additional issue of deep nesting not being a thing you can do if the item doesn't already exist in dynamodb of course. Ideally it would be nice to have a way to do dynamic top level props.

[Playground Link](https://electrodb.fun/?#code/PQKgBAsg9gJgpgGzARwK5wE4Es4GcA0YuccYGeqCALgUQBYCG5YA7llXWAGbZwB2MXGBDAAUKKwBbAA5QMVMAG8wAYVS4qUSQEEqVbACNUVOABUAntLiEAonyrtzhAMqYAblgDGpAL7cMWmAARIhwnvqwBkEA3OKeUHwaYFQMBgikALzB5lCoGAD6KWlw+XwMknAxolSWpABiqHzhWAkAknxuUJ4MDgnOqJKSTOatJpJgWYqiYNwMWOkwAFxgfAMGmLEzuKie3nDwSytrG6I+sdW1YA1NvXztnd23-YPDE0rTYADa0j10yxrYPgAcwAustrs02h0uj0WnxnkMMCMxrEznEEkkVHI4AjXlk+HAWGA7A4agAKKYzSSwRDLSkzMD8UnmZZBeLkfLbF5IoL4D4zNyYXBw1kARl5-KI7i8cFZcEF9g02IlMx8fJmwGAYG6fAA5AouFggXlSLkFHQoETNGAGAgEJawDBzGVJF5klBpGB0oKkOw4JIhLgoGAAAbaAAi4ZDrFyCBgMYwAGsPj19FgjCZcHTJeySnAAB5hYxw-JYQ70hnJWqsgFYYEqytkOBoLDkQ76dDqhlqyVcxEs96NqtWZZqJU6PSGYxmWoAHght3uMKeA37AD4yUEGHxzEEAJRd1VdnszOvwQtZwcMvvDbND6SJu9D7g4OOsh8NofxGRQYUmZafGy2L5AWRa3KWMBBCCh7djBWyPleQ6GoghxBLgiafo236yH+spfNBkpHpKJ5gCRihFOkPinHu5zxIkCgwAYbwEkSrgYB43gUh8ua4rwl5YuQvFONR5yaiswYsAw5juqGxBUNGbAcKGWBcKUUBUCB+ZYBouDRtIAQeDAdZAjajpwFwDCUAo27xhw-A2jARnAkQWhwBwxkyewoiMQAdEy7A4LgPk8auTCBT5qDSDAPRwFxMy5ppYElmWrKpJ4QTUT50UpGSZKphghAenuGRrvSHo+SpABy6k2FpOl5ZOPk3kinz6ngDjArqII+RZ8z7IQAAMNGZUCUBksN4hiQwjnatu+pgOsYDqPsYB1is7UrUMnoIFgiakO5uDeQYfn2AFeDBdivHhZF2WxfSCWgZ4xYJBBqUGOlmXZQwuX5YV0jFaVHzldNMANfoTWhS1bUaMZXU9XMCyEKKw0+HuPmjeNsRAA)

Part of the problem is needing a dynamic map. This is useful for things like tracking aggregates. Any ideas how I could approach this differently with Electro?

**Comment by tywalch** [comment]
Hey @shortjared 👋 

Are you getting an error when you try to perform an `add`? I would suspect (given how deeply nested the structure is) that the error is something about an invalid path? The `ADD` operation (used by ElectroDB) allows for numbers not to exist and will default the values to zero automatically. The challenge is that if the containing `map` (in your case either `summary` or `testing`) does exist then DynamoDB will throw because it doesn't have a map to place the attribute `failed`. If this is the case, DynamoDB doesn't offer much recourse here, but knowing the error you see would help me understand what's happening there.

That said, if the issue is related to the path not existing (no `summary` or `testing` maps prior to your update) here are two ways you might be able to model what you need:

1. [Reduce your nested properties by rethinking the interface a tad bit. Perform an idempotent pre-update and then increment properties off a single map](https://electrodb.fun/?#code/PQKgBAsg9gJgpgGzARwK5wE4Es4GcA0YuccYGeqCALgUQBYCG5YA7llXWAGbZwB2MXGBDAAUKKwBbAA5QMVMAG8wAYVS4qUSQEEqVbACNUVOABUAntLiEAonyrtzYAL7cMWsACJEcAMb7YA08AbnFfKD4NMCoGAwRSAF4vcyhUDAB9GLi4dL4GSTgQ8SpLUgAxVD5-LAiAST4ANyhfBgcIgGVUSUkmJyTFUTAwAG1pVroALiJ9LD4AcwBdKb4ug0xQ51DRcMiFFTk4Tu7esCS+OBYwOwcSgAoBoclYRCmHobB+G-Mpz3DydNwXR6GHMnnwg3eDUwuBqfB+AEYwRChsQMA0sL44D84FD7BoDkihs5wUNgMAwC0+AByBRcLBzNKkVIKOhQS6aMAMBAINlgGDmPKSDHRKDSMDxKFIdhwSRCXBQMAAA20ABEVYrWKkEDBNRgANYQ1ozIwmXCvZEUg7pOAADz8xlh6SwMHN73eJSsPw02HmhLdZDgaCw5Bd0Qw6BJ72JFq4DCw8VDbzdHqxqnUmh0ekMxjMpQAPBUqm0+PUmi1i0dgeYAHy3KkMPjmKkASkjbvgscoVFe0bdvfegN8mLg8ETFqGKamanxmeNOYsVgLlWqdUazVasMrvVr9cbLbb7w7DC7PYPzgh-dm8DtZqUFsBxxBrv90j1z-93Bw2p+r79H-CMhQDCJhTMMvxWra9rFk6MCeAsB5EghRBvneH5DHSiChp4uB6n+-oAbIwGpsM8Hji4Z6RtGxJKFk8TOM2WxkpyCgcKQGhMAoUBcGAKRpGAsxlhuESEEeXY8akGBgA0XLoEItwcFgQiKfx8CASY9jNtsERRH8OSQb4DoRDBpxgPWBi+FSoT7OQW4ggAdKg0gwK0cD3Lp1p2gZ0HOvRdnOTEty3EaGCEKKzanNWqFgKKdlYFwAByUBUDYNqKTQQVZnZsbxiOhCKPRoRDDFcWJclqUaLgGX6HZg7DqOeUFaIvlzFAtwMeITEsHQ-DiagYDnCOIr8VU5AFPYojWYcQK9A5TkuW5EGeYZfAwb5-kMIFwWhdI4UJJFbwxQwMAwFVGBZXGCZ2aaDjzIQ8LtUV0h2UdJ3BTVqBDiQo5XXgN1zHd7XNa1DFAA)

2. [Isolate your summary types with entities to simply updates and use collections to simply queries](https://electrodb.fun/?ssl=73&ssc=69&pln=73&pc=3#code/PQKgBAsg9gJgpgGzARwK5wE4Es4GcA0YuccYGeqCALgUQBYCG5YA7llXWAGbZwB2MXGBDAAUKKwBbAA5QMVMAG8wAYVS4qUSQEEqVbACNUVOABUAntLiEAonyrtzhAMqYAblgDGpAL7cMWmAARIhwnvqwBkEA3OKeUHwaYFQMBgikALzB5lCoGAD6KWlw+XwMknAx4lSWpABiqHzhWAkAknxuUJ4MDgnOqJKSTOatJpJgWYqiYNwMWOkwAFxgfAMGmLEzuKie3nDwSytrG6I+saI1VmANTb187Z3dd-2DwxNK02AA2tI9dMsabB8ADmAF1ljdmm0Ol0ei0+C8hhgRmNYmdxFxGlC+GBPOQenAVHI4IjhnYHDUABRlCoA-RYEEASg+M3IVDyOL4cBYYHJjkpUxmM0ksEQy0FQpm-Ap5mWAAN4uR8ttXsj8gASRQ0uA+OX4T6StyYXDw5ZBACMQX1kq27i8cDNcCN9g0xKtBrAPmtQuAwFxDD4AHIFFwsMC8qRcgo6FAeZowAwEAhY2AYOYaV5klBpGB0kakOw4JIhLgoGA5doACKVuWsXIIGB1jAAaw9PXpRhMuHFHpmipKcAAHmFjPD8lhDhKbclambAQzge7p6y4GgsORDvp0N7JV7e7N5vse8uZ1ZloHVpJ1hhAzubeQ1xvllvrPuZvAuAxKFRlgAGO8zHu07bLsJAHMey6XA6YAXscN4AUKD6oOuR7JBg25vqmcCft+f4IT4HpAUKDLwMO3YsjaKpIrKFHTvESZhHcyxfIGVFMDguCBqCCEzNIzYQSeoaIIcQR8UuJ64losgmiYzFBP2+RDiOdzjjAQTcZhRHAfxtHLkJDZzs24knvEMhQDJ0FfBpJ5aUKtlEV6ShFOkPiMmicQJEk-aksiph4A4ILvHicAEkS5A+eYfJUsG-kLoGbmiPEiQKN5AzUQA8hwmBBfiJhhSSaVkvY-KBlAWU3gloi+mAWClggBJEIVyKnngCYKFlYDSo4uZOogiWeQoMAGO8XI8q4GAeN4AqfKlqrmH5GgLt6s0ZeV+quecSVecSinDp4o4JKp7yBqkniBuc1UMDAjZsBwUYJnw5iSU0mCJNwchYdIHA1bg2x4P14VNfNsUggAdKg0gwASAoKUp+0qROrmg1dMACiBezgWA5pI8CUCUglK3DJldCYODkPQ4osN7QdfCqUjKMCp+h6HNjjKg7j+MXX66ikPR6TYkI8ZoJg5gAAWJkgbG8EIt2cAwRALukZCrugGiiENoN84x8K4KDUscTDO1wzTdNsxzbNZXwlIClDKSuRMAB8FHVSaMhK-oAa4FwchDHcWZgMCcDtWWkJ3A8sLPED9AMFYM0DY1c0QliYcwk88IRe8ijojM3sYGAlJbQoUvzbUNU4rbDDMlOz1JF8biJqroLvBXXzFxYVwMEIzZwDkXAtVAfcV6CmxCsXrdA+3cBN1k9cIKrI8ESu7IYDixdoglQA)

Let me know about the error message you see, whether what you're seeing is an error path issue, and/or if any of the alternative models I put above work for your use case 👍

**Comment by shortjared** [comment]
You are correct in the not having the property ahead of time being part of the issue with regards to `ADD` on the update operation. That's a solvable problem.

Part of the issue is I don't know the names of the properties ahead of time. If Electro had a way to allow "dynamic" patterned top level properties, I could arbitrarily use `ADD` to track such aggregates. I can't seem to figure out a reasonable way to do it. (I can of course eject to just normal document client, but I lose a lot of the nice bits of using Electro)

```
core_execution_id: abc
2023-04-28: 198
2023-04-29: 203
2023-04-30: 101
```

I know this isn't a core concern of the lib, so no worries if there isn't a clean way to figure it out!

**Reply by tywalch** [reply]
Yup, having explicit top-level attributes is a pretty static constraint currently. 

Is your dynamic property a date? If so, have you considering simply making a record for each execution? That would work well sort keys: [playground](https://electrodb.fun/?#code/JYWwDg9gTgLgBAbzgUQHY2DAnnAvnAMyghDgCIBTAGwoGMZiATAIzIG4AoD2iVAZ3gwAhsxpwAvOSwQArlAD6w0RXmohICuy49+8HlAoBlGSBBCoOSagoB3FOkxYAFAg5w4ICI2oAuRG-c4CgdsPwADfRU+EzMLMIAaAPcANwooPmBePzIARjJEwLg+NOTgWgpsilT0AWhNArwGgHomuFohVAByeAJgAHM5CjhZeAALCDsYCDghKioJuEYsNRAyuCmwOBpUqjhMChA+IumwgEEAEXOwuBtZKkYb6ABrAKEYBmBmGRgKPj9XQqReQUAAedG+mVQ8mAjH+SUC2DAFXIAigwFQfXy8PcBgAjjJgAZYesoDIKA13LgKYs3siAYV1lgkX5Oqj0X1OvCqfCCEJgDRifTCojkZ1UCZmGlOtScRR8YSKMSGGSZYsKLyZFQYH4AAzU7mFaK0cqKxVwhmM5lwMUSqWqvEEol+ZXk7FqjVa3X6gIGvaobxgv7+QrMLDnWnmhlgJ6Ri29ajEsjRrEW9w8cAQDI-PwAbTIQNB4IwvGhjDIAF1Vb7DTHg6n4-dsnwninU+nIFnkXnGLSK1X9Q1uVTEEoaLgAJScDgEGSoeiQuB9CgwAAqXiEznHdYMMDkqDg1js4Z+TnHADopgBJQwAeUMHwxp7PfDAVEwTk6K8645zOvLnFwbReAENo6mBMFaAhEsYQka0RFoTpOB0ECex+WCl1XddNynFoZkYB4bEwcZvhmVAcB0cooH4QhoDVMAYFGPY+GiX5uDqYxTHMLAAjPGQwFQigXFAgxwKLSFS3iGk0InHihHwlwjRNbxiRyGT3DPPoIFPHDWgw4ShmiTi0V+OBQykoYhD4cp-XZNiDA42IsDPfE0ic0Nj0EhACwgqCoRhCcNK0ycuFwvTIiKGJzGAEyzIEtUrOCRhbMiByuOcskLDPdzaRcbyxOgxgAs0lxoG8KAWW8KzOgnHTF2XfSIqM6KjgIWiGKGLAKHMYYCDgAAmHU+oAZjsoxIsylyLB47KTy8sDC0g4s-MK8dpooPp0T4FwBJZAbhuq1b1OK4KOFC+rwsM2JmpoqA4AARRyfrBpGlLxqcybuPUmbPLyxbxP8w64Cy5cbAoYInHhBAdutPahoAWh1HIEZyarqShiMYeehH4aGlHfUBwLtJC3TzrqRrHJuookVoYBeloGluNeoynIwoSAl+3yJLgAJoYwtce03RIiqCtggA).

If it's truely dynamic (you don't know the value ahead of time) you could also simply add it to your key structure: [playground](https://electrodb.fun/?#code/JYWwDg9gTgLgBAbzgUQHY2DAnnAvnAMyghDgCIBTAGwoGMZiATAIzIG4AoD2iVAZ3gwAhsxpwAvOSwQArlAD6w0RXmohICuy49+8HlAoBlGSBBCoOSagoB3FOkxYAFAg5w4ICI2oAuRG-c4CgdsPwADfRU+EzMLMIAaAPcANwooPmBePzIARjJEwLg+NOTgWgpsilT0AWhNArwGgHomuFohVAByeAJgAHM5CjhZeAALCDsYCDghKioJuEYsNRAyuCmwOBpUqjhMChA+IumwgEEAEXOwuBtZKkYb6ABrAKEYBmBmGRgKPj9XQqReQUAAedG+mVQ8mAjH+SUC2DAFXIAigwFQfXy8PcBgAjjJgAZYesoDIKA13LgKUUYuYsAAVLBIuGFdyI5GdVHovqdeFU+EEITAGjEgGs9l+TqoEzMNKdak4ij4wkUYkMMkKxYUQUyKgwPwABmp-MK0Vo5VVqpZ4qZHOlIFlUHl2LgeIJRL86vJLu8Or1huNARNe1Q3jBf38hWYWHOb2RYsKYCe1tZhGA1GJZCTWNT7h44AgGR+fgA2mQgaDwRheNDGGQALqa4Om5OR3O9DPZPhPHO5-OQIvIsvRUx0xlIhtN40NflUxBKGi4ACUnA4BBkqHokLgfQoMHpXiEziXbYMMDkqDg1jssZ+TiXADopgBJQwAeUMHwx94ffDAVEwJxOnpTolxLA1604XBtF4AQ2jqYEwVoCEaxhCQ4E6ERaE6TgdDgkdYgZW10O6X4MAxXkOBaGZGAeGxMHGb4ZlQHAdHKKB+EIaAtTAGBRj2Phol+bg6mMUcLACB8ZDARg4xceCDEQqtIVreIaXEoikTwJdJKEWiXDNC1vGJHJl0kvoIHvNggA).

I should have suggested this from the start

**Reply by tywalch** [reply]
You might also try to define maps you need when provisioning the dynamodb table 👍

**Reply by shortjared** [reply]
Many records in this case is painful because it's a unique need where I may have 10-100 different aggregates on the same record I could update at the same time. It becomes a cost and performance optimization to do everything in a singular write instead of many writes.

Like I said, totally realize I'm trying to get electro to do something it isn't designed for! Thanks for your help in any case, great lib!
